dao/job_metadata.go

```diff
@@ -0,0 +1,156 @@
+			SQL: `INSERT INTO SMT_JOB 
+			(JobId, JobName, JobType, JobStateData, JobData, Dialect, SpannerDatabaseName, CreatedAt,UpdatedAt)
+			VALUES(
+			 @jobId, @jobName, @jobType, @jobStateData, @jobData, @dialect, @dbName, PENDING_COMMIT_TIMESTAMP(), PENDING_COMMIT_TIMESTAMP()
+			);`,
+			Params: map[string]interface{}{
+				"jobId":        jobId,
+				"jobName":      jobName,
+				"jobType":      jobType,
+				"jobStateData": spanner.NullJSON{Valid: true, Value: StateData{State: "CREATING"}},
+				"jobData":      jobData,
+				"dialect":      dialect,
+				"dbName":       dbName,
+			},
+		}
+		_, err := txn.Update(ctx, jobStmt)
+		if err != nil {
+			return err
+		}
+		// Update job history table within the same txn.
+		_, err = updateJobHistoryWithinTxn(ctx, txn, jobId)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("could not insert SMT job entry: %v", err)
+	}
+	return nil
+}
+
+func updateJobHistoryWithinTxn(ctx context.Context, txn *spanner.ReadWriteTransaction, jobId string) (int64, error) {
+	// Fetch the latest version for job.
+	version, err := getLatestJobVersionWithinTxn(ctx, txn, jobId)
+	if err != nil {
+		return 0, fmt.Errorf("error fetching latest job version: %v", err)
+	}
+	// Fetch the current row from SMT_JOB table.
+	stmt := spanner.Statement{SQL: `
+		SELECT 
+			JobName, JobType, JobStateData, JobData, Dialect, SpannerDatabaseName 
+		FROM SMT_JOB WHERE JobId = @jobId;`,
+		Params: map[string]interface{}{"jobId": jobId},
+	}
+	iter := txn.Query(ctx, stmt)
+	defer iter.Stop()
+	var jobName, jobType, dialect, spannerDatabaseName spanner.NullString
+	var jobStateData, jobData spanner.NullJSON
+	row, err := iter.Next()
+	if err == iterator.Done || err != nil {
+		return 0, err
+	}
+	if err := row.Columns(&jobName, &jobType, &jobStateData, &jobData, &dialect, &spannerDatabaseName); err != nil {
+		return 0, fmt.Errorf("error reading smt job row: %v", err)
+	}
+
+	// Insert entry to SMT_JOB_HISTORY table.
+	jobStmt := spanner.Statement{
+		SQL: `INSERT INTO SMT_JOB_HISTORY 
+		(JobId, Version, JobName, JobType, JobStateData, JobData, Dialect, SpannerDatabaseName, CreatedAt)
+		VALUES(
+		 @jobId, @version, @jobName, @jobType, @jobStateData, @jobData, @dialect, @spannerDatabaseName, PENDING_COMMIT_TIMESTAMP()
+		);`,
+		Params: map[string]interface{}{
+			"jobId":               jobId,
+			"version":             version + 1,
+			"jobName":             jobName,
+			"jobType":             jobType,
+			"jobStateData":        jobStateData,
+			"jobData":             jobData,
+			"dialect":             dialect,
+			"spannerDatabaseName": spannerDatabaseName,
+		},
+	}
+	return txn.Update(ctx, jobStmt)
+}
+
+func getLatestJobVersionWithinTxn(ctx context.Context, txn *spanner.ReadWriteTransaction, jobId string) (int64, error) {
+	// Fetch rows for the job with latest version.
+	stmt := spanner.Statement{SQL: `SELECT MAX(Version) FROM SMT_JOB_HISTORY WHERE JobId = @jobId;`,
```

Comment:
Wouldn't the latest version be the one in SMT_JOB table?

Reply:
This is fetching the latest version number, which is in the history table. Updated the comment to be more accurate.